### Теорія

Геометрична прогресія

$Sn=b1 * q^n−1$, де 

`n`- порядковий номер члена прогресії,

`b1`- перший член послідовності,

`q`- знаменник.

Сума геометричної прогресії

$Sn = b1 · 	(1 - q^n)/ 1 - q$,  де 

`n`- кількість членів послідовності (порядковий номер),

`b1`- перший член послідовності,

`Sn`- n-ий член послідовності, 

`q`- знаменник.  
 
Згенерований дохід

$Rtotal = price * S1 * (1 - q**n) / 1 - q$

`price `= 100

`S1(subscribers)`= 500

`q `= 0.9

`n(period)` = 6

`Rtotal = 100 * 500 * (1 - 0.9**6) / (1 - 0.9) = 234 * 10**3`



### Задачі

#### Задача 1

Ви продакт-менеджер у компанії, яка займається мобільними додатками. Ваша
модель монетизацї – це підписка.
У травні 2019 року до вас приєдналося 100 підписникв, у червні – 150, у липні –
170. Скільки ці користувачі принесли вам до моменту
Травень 2021 року, якщо щомісячний відтік становить 5%, а підписка коштує 10
доларів США?


In [85]:
def calc_revenue(price, churn_rate, total_months, subscribers):
    churn_rate = 1 - (churn_rate / 100)
    total = price * subscribers * (1 - churn_rate**total_months)/(1 - churn_rate)
    return round(total,2)

In [86]:
price = 10
churn_rate = 5
total_months = 24
#загальна кількість підписників
subscribers = 420

In [87]:
result = calc_revenue(price, churn_rate, total_months, subscribers)
print(f"Загальний дохід: ${result}")

Загальний дохід: $59472.92


У цій задачі ми використовуємо суму геометричної прогресії $Sn = b1 * (1 - q^n)/ 1 - q$, для знаходження загального прибутку за період 24 місяці. Так як підписка зменшується кожен місяць на 5% отримаємо коефіцєнт 0,95 та возведемо у ступінь нашого періоду. Отримаємо такий вираз:

`Rtotal = 10 * 420 * (1 - 0.95^24)/(1 - 0.95)`

Порахуємо те що в дужках, потім верхню частину, а потім розділимо на результат нашого знаменника. Округлимо до сотих.

Загальний дохід: $59472.92

#### Задача 2

Ви — менеджер по роботі з продуктом в компанії, яка розробляє мобільні ігри. Вашою метою є збільшення кількості активних користувачів. У січні 2020 року у вас було 2000 активних користувачів. З лютого кількість активних користувачів щомісяця зростає на 3%. У той же час, 2% користувачів щомісяця перестають користуватися додатком. Якою буде кількість активних користувачів у грудні 2021 року?

In [107]:
subscribers =2000
churn_rate = 1
total_months = 24

In [111]:
def total_subscribers( churn_rate, subscribers, total_months=1):
    churn_rate = churn_rate / 100
    # Sn=b1 * q**n−1
    if total_months > 1:
        total = subscribers * ((1 + churn_rate)**(total_months - 1))
    else:
        total = subscribers * ((1 + churn_rate)**total_months)
    return round(total)

In [109]:
result = total_subscribers(churn_rate, subscribers, total_months, )
print(f"Активні користувачі: {result}")

Активні користувачі: 2514


За геометричною прогресією $Sn=s1 * q^n−1$ знаходимо кількість користувачів у грудні 2021 році. Кожного місяця додається 1% користувачів, а за весь період отримаємо коефіцієнт 1,25. Помножемо на початкове значення користувачів та отримаємо відповідь 2514 користувачів.

#### Задача 3

Ви — менеджер з розвитку бізнесу в компанії, що займається створенням платформи для онлайн-курсів. У січні 2021 року ви залучили 500 нових користувачів. У лютому кількість нових користувачів зросла на 50%, у березні — ще на 20%. З квітня приріст нових користувачів стабілізувався на рівні 10% щомісяця. Скільки нових користувачів ви залучили загалом за перше півріччя 2021 року?

In [120]:
subscribers =500
churn_rate_1 = 50
churn_rate_2 = 20
churn_rate_3 = 10
total_months = 3

In [124]:
feb_users = total_subscribers(churn_rate_1, subscribers)
march_users = total_subscribers(churn_rate_2, feb_users)
other_users = total_subscribers(churn_rate_3, march_users, total_months)
display(feb_users,march_users,other_users)

750

900

1089

У третьому завданні обчислень більше за рахунок різного коефіцієнту. Спочатку нам потрібно обчислити лютий місяць +50%, потім березень +20%, а потім вже за геометричною прогресією обчислюємо останні три місяці по 10%. У перших двох обчисленнях ми не можемо скористатись геометричною прогресією, так як `n` з формули $Sn=s1 * q^n−1$ дорівнює одиниці,це повертає нам нульову ступінь, тобто коефіцієнт буде дорівнювати одиниці. Тому підправив трошки функцію total_subscribers. Результат 1089 користувачів за перше півріччя 2021 року.